IllegalStateException in TabbedViewPager

### Describe your problem!

This is reported by GooglePlayConsole as new with current beta 2022.05.02-RC:

```
java.lang.IllegalStateException: 
  at androidx.fragment.app.FragmentManager.ensureExecReady (FragmentManager.java:1931)
  at androidx.fragment.app.FragmentManager.execSingleAction (FragmentManager.java:1967)
  at androidx.fragment.app.BackStackRecord.commitNow (BackStackRecord.java:305)
  at androidx.viewpager2.adapter.FragmentStateAdapter$FragmentMaxLifecycleEnforcer.updateFragmentMaxLifecycle (FragmentStateAdapter.java:726)
  at androidx.viewpager2.adapter.FragmentStateAdapter$FragmentMaxLifecycleEnforcer$2.onChanged (FragmentStateAdapter.java:646)
  at androidx.recyclerview.widget.RecyclerView$AdapterDataObservable.notifyChanged (RecyclerView.java:12653)
  at androidx.recyclerview.widget.RecyclerView$Adapter.notifyDataSetChanged (RecyclerView.java:7584)
  at cgeo.geocaching.activity.TabbedViewPagerActivity.notifyAdapterDataSetChanged (TabbedViewPagerActivity.java:195)
  at cgeo.geocaching.activity.TabbedViewPagerActivity.setOrderedPages (TabbedViewPagerActivity.java:112)
  at cgeo.geocaching.CacheDetailActivity.notifyDataSetChanged (CacheDetailActivity.java:930)
  at cgeo.geocaching.CacheDetailActivity.ensureSaved (CacheDetailActivity.java:1999)
  at cgeo.geocaching.VariablesViewPageFragment.checkUnsavedChanges (VariablesViewPageFragment.java:123)
  at cgeo.geocaching.VariablesViewPageFragment.onPause (VariablesViewPageFragment.java:158)
  at androidx.fragment.app.Fragment.performPause (Fragment.java:3168)
  at androidx.fragment.app.FragmentStateManager.pause (FragmentStateManager.java:632)
  at androidx.fragment.app.FragmentStateManager.moveToExpectedState (FragmentStateManager.java:314)
  at androidx.fragment.app.FragmentManager.executeOpsTogether (FragmentManager.java:2189)
  at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute (FragmentManager.java:2100)
  at androidx.fragment.app.FragmentManager.execSingleAction (FragmentManager.java:1971)
  at androidx.fragment.app.BackStackRecord.commitNow (BackStackRecord.java:305)
  at androidx.viewpager2.adapter.FragmentStateAdapter$FragmentMaxLifecycleEnforcer.updateFragmentMaxLifecycle (FragmentStateAdapter.java:726)
  at androidx.viewpager2.adapter.FragmentStateAdapter$FragmentMaxLifecycleEnforcer$1.onPageScrollStateChanged (FragmentStateAdapter.java:632)
  at androidx.viewpager2.widget.CompositeOnPageChangeCallback.onPageScrollStateChanged (CompositeOnPageChangeCallback.java:87)
  at androidx.viewpager2.widget.CompositeOnPageChangeCallback.onPageScrollStateChanged (CompositeOnPageChangeCallback.java:87)
  at androidx.viewpager2.widget.ScrollEventAdapter.dispatchStateChanged (ScrollEventAdapter.java:426)
  at androidx.viewpager2.widget.ScrollEventAdapter.onScrolled (ScrollEventAdapter.java:214)
  at androidx.recyclerview.widget.RecyclerView.dispatchOnScrolled (RecyclerView.java:5347)
  at androidx.recyclerview.widget.RecyclerView$ViewFlinger.run (RecyclerView.java:5512)
  at android.view.Choreographer$CallbackRecord.run (Choreographer.java:1108)
  at android.view.Choreographer.doCallbacks (Choreographer.java:866)
  at android.view.Choreographer.doFrame (Choreographer.java:792)
  at android.view.Choreographer$FrameDisplayEventReceiver.run (Choreographer.java:1092)
  at android.os.Handler.handleCallback (Handler.java:938)
  at android.os.Handler.dispatchMessage (Handler.java:99)
  at android.os.Looper.loopOnce (Looper.java:226)
  at android.os.Looper.loop (Looper.java:313)
  at android.app.ActivityThread.main (ActivityThread.java:8663)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:567)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1135)
```


### How to reproduce?

unclear

### Actual result after these steps?

_No response_

### Expected result after these steps?

_No response_

### Reproducible

Unclear

### c:geo Version

2022.05.02-RC2

### System information

```text
Android 9 - 12
25 occurrences, 19 users
```


### Additional Information

_No response_

Similar:
```
java.lang.RuntimeException: 
  at android.app.ActivityThread.performPauseActivityIfNeeded (ActivityThread.java:5557)
  at android.app.ActivityThread.performPauseActivity (ActivityThread.java:5508)
  at android.app.ActivityThread.handlePauseActivity (ActivityThread.java:5460)
  at android.app.servertransaction.PauseActivityItem.execute (PauseActivityItem.java:47)
  at android.app.servertransaction.ActivityTransactionItem.execute (ActivityTransactionItem.java:45)
  at android.app.servertransaction.TransactionExecutor.executeLifecycleState (TransactionExecutor.java:176)
  at android.app.servertransaction.TransactionExecutor.execute (TransactionExecutor.java:97)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:2438)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loopOnce (Looper.java:226)
  at android.os.Looper.loop (Looper.java:313)
  at android.app.ActivityThread.main (ActivityThread.java:8663)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:567)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1135)
Caused by: java.lang.IllegalStateException: 
  at androidx.fragment.app.FragmentManager.ensureExecReady (FragmentManager.java:1931)
  at androidx.fragment.app.FragmentManager.execSingleAction (FragmentManager.java:1967)
  at androidx.fragment.app.BackStackRecord.commitNow (BackStackRecord.java:305)
  at androidx.viewpager2.adapter.FragmentStateAdapter$FragmentMaxLifecycleEnforcer.updateFragmentMaxLifecycle (FragmentStateAdapter.java:726)
  at androidx.viewpager2.adapter.FragmentStateAdapter$FragmentMaxLifecycleEnforcer$2.onChanged (FragmentStateAdapter.java:646)
  at androidx.recyclerview.widget.RecyclerView$AdapterDataObservable.notifyChanged (RecyclerView.java:12653)
  at androidx.recyclerview.widget.RecyclerView$Adapter.notifyDataSetChanged (RecyclerView.java:7584)
  at cgeo.geocaching.activity.TabbedViewPagerActivity.notifyAdapterDataSetChanged (TabbedViewPagerActivity.java:195)
  at cgeo.geocaching.activity.TabbedViewPagerActivity.setOrderedPages (TabbedViewPagerActivity.java:112)
  at cgeo.geocaching.CacheDetailActivity.notifyDataSetChanged (CacheDetailActivity.java:930)
  at cgeo.geocaching.CacheDetailActivity.ensureSaved (CacheDetailActivity.java:1999)
  at cgeo.geocaching.VariablesViewPageFragment.checkUnsavedChanges (VariablesViewPageFragment.java:123)
  at cgeo.geocaching.VariablesViewPageFragment.onPause (VariablesViewPageFragment.java:158)
  at androidx.fragment.app.Fragment.performPause (Fragment.java:3168)
  at androidx.fragment.app.FragmentStateManager.pause (FragmentStateManager.java:632)
  at androidx.fragment.app.FragmentStateManager.moveToExpectedState (FragmentStateManager.java:314)
  at androidx.fragment.app.FragmentStore.moveToExpectedState (FragmentStore.java:112)
  at androidx.fragment.app.FragmentManager.moveToState (FragmentManager.java:1647)
  at androidx.fragment.app.FragmentManager.dispatchStateChange (FragmentManager.java:3128)
  at androidx.fragment.app.FragmentManager.dispatchPause (FragmentManager.java:3090)
  at androidx.fragment.app.FragmentController.dispatchPause (FragmentController.java:284)
  at androidx.fragment.app.FragmentActivity.onPause (FragmentActivity.java:390)
  at cgeo.geocaching.activity.AbstractActivity.onPause (AbstractActivity.java:118)
  at cgeo.geocaching.CacheDetailActivity.onPause (CacheDetailActivity.java:443)
  at android.app.Activity.performPause (Activity.java:8485)
  at android.app.Instrumentation.callActivityOnPause (Instrumentation.java:1530)
  at android.app.ActivityThread.performPauseActivityIfNeeded (ActivityThread.java:5547)
```

Similar:

```
java.lang.RuntimeException: 
  at android.app.ActivityThread.performPauseActivityIfNeeded (ActivityThread.java:4750)
  at android.app.ActivityThread.performPauseActivity (ActivityThread.java:4701)
  at android.app.ActivityThread.handlePauseActivity (ActivityThread.java:4652)
  at android.app.servertransaction.PauseActivityItem.execute (PauseActivityItem.java:46)
  at android.app.servertransaction.TransactionExecutor.executeLifecycleState (TransactionExecutor.java:176)
  at android.app.servertransaction.TransactionExecutor.execute (TransactionExecutor.java:97)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:2071)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loop (Looper.java:368)
  at android.app.ActivityThread.main (ActivityThread.java:7710)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:592)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:954)
Caused by: java.lang.IllegalStateException: 
  at androidx.fragment.app.FragmentManager.ensureExecReady (FragmentManager.java:1931)
  at androidx.fragment.app.FragmentManager.execSingleAction (FragmentManager.java:1967)
  at androidx.fragment.app.BackStackRecord.commitNow (BackStackRecord.java:305)
  at androidx.viewpager2.adapter.FragmentStateAdapter$FragmentMaxLifecycleEnforcer.updateFragmentMaxLifecycle (FragmentStateAdapter.java:726)
  at androidx.viewpager2.adapter.FragmentStateAdapter$FragmentMaxLifecycleEnforcer$2.onChanged (FragmentStateAdapter.java:646)
  at androidx.recyclerview.widget.RecyclerView$AdapterDataObservable.notifyChanged (RecyclerView.java:12653)
  at androidx.recyclerview.widget.RecyclerView$Adapter.notifyDataSetChanged (RecyclerView.java:7584)
  at cgeo.geocaching.activity.TabbedViewPagerActivity.notifyAdapterDataSetChanged (TabbedViewPagerActivity.java:195)
  at cgeo.geocaching.activity.TabbedViewPagerActivity.reinitializePage (TabbedViewPagerActivity.java:219)
  at cgeo.geocaching.VariablesViewPageFragment.lambda$createView$0 (VariablesViewPageFragment.java:47)
  at cgeo.geocaching.VariablesViewPageFragment.lambda$createView$0$VariablesViewPageFragment (Unknown Source)
  at cgeo.geocaching.-$$Lambda$VariablesViewPageFragment$ElyYPa69Le2ifPebGt211wwhpQE.accept (Unknown Source:6)
  at cgeo.geocaching.ui.VariableListView$VariablesListAdapter.callCallback (VariableListView.java:546)
  at cgeo.geocaching.ui.VariableListView$VariablesListAdapter.setVariableList (VariableListView.java:221)
  at cgeo.geocaching.VariablesViewPageFragment.setContent (VariablesViewPageFragment.java:115)
  at cgeo.geocaching.activity.TabbedViewPagerFragment.notifyDataSetChanged (TabbedViewPagerFragment.java:61)
  at cgeo.geocaching.activity.TabbedViewPagerActivity.setOrderedPages (TabbedViewPagerActivity.java:110)
  at cgeo.geocaching.CacheDetailActivity.notifyDataSetChanged (CacheDetailActivity.java:930)
  at cgeo.geocaching.CacheDetailActivity.ensureSaved (CacheDetailActivity.java:1999)
  at cgeo.geocaching.VariablesViewPageFragment.checkUnsavedChanges (VariablesViewPageFragment.java:123)
  at cgeo.geocaching.VariablesViewPageFragment.setContent (VariablesViewPageFragment.java:104)
  at cgeo.geocaching.activity.TabbedViewPagerFragment.notifyDataSetChanged (TabbedViewPagerFragment.java:61)
  at cgeo.geocaching.activity.TabbedViewPagerActivity.setOrderedPages (TabbedViewPagerActivity.java:110)
  at cgeo.geocaching.CacheDetailActivity.notifyDataSetChanged (CacheDetailActivity.java:930)
  at cgeo.geocaching.CacheDetailActivity.ensureSaved (CacheDetailActivity.java:1999)
  at cgeo.geocaching.VariablesViewPageFragment.checkUnsavedChanges (VariablesViewPageFragment.java:123)
  at cgeo.geocaching.VariablesViewPageFragment.onPause (VariablesViewPageFragment.java:158)
  at androidx.fragment.app.Fragment.performPause (Fragment.java:3168)
  at androidx.fragment.app.FragmentStateManager.pause (FragmentStateManager.java:632)
  at androidx.fragment.app.FragmentStateManager.moveToExpectedState (FragmentStateManager.java:314)
  at androidx.fragment.app.FragmentStore.moveToExpectedState (FragmentStore.java:112)
  at androidx.fragment.app.FragmentManager.moveToState (FragmentManager.java:1647)
  at androidx.fragment.app.FragmentManager.dispatchStateChange (FragmentManager.java:3128)
  at androidx.fragment.app.FragmentManager.dispatchPause (FragmentManager.java:3090)
  at androidx.fragment.app.FragmentController.dispatchPause (FragmentController.java:284)
  at androidx.fragment.app.FragmentActivity.onPause (FragmentActivity.java:390)
  at cgeo.geocaching.activity.AbstractActivity.onPause (AbstractActivity.java:118)
  at cgeo.geocaching.CacheDetailActivity.onPause (CacheDetailActivity.java:443)
  at android.app.Activity.performPause (Activity.java:8174)
  at android.app.Instrumentation.callActivityOnPause (Instrumentation.java:1511)
  at android.app.ActivityThread.performPauseActivityIfNeeded (ActivityThread.java:4740)
```

--- a/main/src/cgeo/geocaching/VariablesViewPageFragment.java
+++ b/main/src/cgeo/geocaching/VariablesViewPageFragment.java
@@ -120,8 +120,11 @@ public class VariablesViewPageFragment extends TabbedViewPagerFragment<Cachedeta
 
     private void checkUnsavedChanges() {
         if (cache != null && cache.getVariables() != null && cache.getVariables().wasModified()) {
-            activity.ensureSaved();
-            cache.getVariables().saveState();
+            //make call asynchronous due to possible fragment transaction conflicts, see #12977
+            binding.getRoot().post(() -> {
+                activity.ensureSaved();
+                cache.getVariables().saveState();
+           });
         }
     }
 

--- a/main/src/cgeo/geocaching/calculator/CalculatedCoordinateMigrator.java
+++ b/main/src/cgeo/geocaching/calculator/CalculatedCoordinateMigrator.java
@@ -284,7 +284,12 @@ public class CalculatedCoordinateMigrator {
     public CalculatedCoordinateMigrator(final Map<String, String> initialVars, final WaypointMigrationData wmd) {
         this.initialVars = initialVars;
         waypointMigrationData = wmd;
-        final Map<String, String> cacheVars = new HashMap<>(initialVars);
+        final Map<String, String> cacheVars = new HashMap<>();
+        for (Map.Entry<String, String> iv : initialVars.entrySet()) {
+            if (iv.getKey() != null && iv.getValue() != null) {
+                cacheVars.put(iv.getKey(), iv.getValue());
+            }
+        }
 
         //find out what needs to be replaced and do it
         final Map<String, String> replacements = new HashMap<>();

--- a/main/src/cgeo/geocaching/ui/VariableListView.java
+++ b/main/src/cgeo/geocaching/ui/VariableListView.java
@@ -124,14 +124,14 @@ public class VariableListView extends LinearLayout {
 
         private void setData(final VariableMap.VariableState variableState) {
 
-            this.varName = variableState.getVar();
+            this.varName = variableState == null ? "?" : variableState.getVar();
             final String displayVarName = VariableList.isVisible(this.varName) ? this.varName : "-";
             this.displayType.setVariableName(this.viewVariableName, displayVarName);
 
             if (this.viewVariableFormulaText != null) {
                 final String currentText = this.viewVariableFormulaText.getText().toString();
-                if (!currentText.equals(variableState.getFormulaString())) {
-                    this.viewVariableFormulaText.setText(variableState.getFormulaString());
+                if (!currentText.equals(variableState == null ? "" : variableState.getFormulaString())) {
+                    this.viewVariableFormulaText.setText(variableState == null ? "" : variableState.getFormulaString());
                 }
                 if (listAdapter.currentFocusKeep && Objects.equals(listAdapter.currentFocusVar, this.varName)) {
                     // we come here if listview changed due to added view (due to filter being changed)
@@ -155,10 +155,10 @@ public class VariableListView extends LinearLayout {
             if (viewVariableResult == null) {
                 return;
             }
-            final boolean isError = variableState.getError() != null;
+            final boolean isError = variableState == null || variableState.getError() != null;
             if (isError) {
-                final CharSequence errorText = TextUtils.setSpan(variableState.getError(), new ForegroundColorSpan(Color.RED));
-                if (variableState.getResultAsCharSequence() != null) {
+                final CharSequence errorText = variableState == null ? "?" : TextUtils.setSpan(variableState.getError(), new ForegroundColorSpan(Color.RED));
+                if (variableState != null && variableState.getResultAsCharSequence() != null) {
                     viewVariableResult.setText(TextUtils.concat(variableState.getResultAsCharSequence(), " | ", errorText));
                 } else {
                     viewVariableResult.setText(errorText);

